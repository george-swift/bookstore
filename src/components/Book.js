import PropTypes from 'prop-types';
import { CircularProgressbar } from 'react-circular-progressbar';
import 'react-circular-progressbar/dist/styles.css';
import { Link } from 'react-router-dom';

const Book = ({ book, editBook, removeBook }) => (
  <div className="col-12 d-flex book-row">
    <div className="book-name">
      <div className="details">
        <p className="book-category">{book.category}</p>
        <p className="title">{book.title}</p>
        <span className="d-block author">{book.author || 'Click Edit to add author'}</span>
        <ul>
          <li className="action border-end"><span className="comments" role="button">Comments</span></li>
          <li className="action border-end">
            <button
              type="button"
              className="remove"
              onClick={() => removeBook(book)}
              title="Click to remove book from store"
            >
              Remove
            </button>
          </li>
          <li className="action">
            <Link to="/edit" className="edit" onClick={() => editBook(book)}>Edit</Link>
          </li>
        </ul>
      </div>
      <div className="progress-ui">
        <span className="progress-icon">
          <CircularProgressbar value={book.percentage || 0} />
        </span>
        <p>
          <span className="percent">{`${book.percentage || 0}%`}</span>
          <br />
          <span className="completed">Completed</span>
        </p>
      </div>
    </div>
    <span className="divider" />
    <div className="book-chapter">
      <p className="current-chapter">Current Chapter</p>
      <p className="current-lesson">{`Chapter ${book.chapter || 0}`}</p>
      <Link to="/edit" className="btn btn-sm update-progress" onClick={() => editBook(book)}>
        Update Progress
      </Link>
    </div>
  </div>
);

Book.propTypes = {
  book: PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    category: PropTypes.string.isRequired,
    author: PropTypes.string,
    percentage: PropTypes.number,
    chapter: PropTypes.number,
  }).isRequired,
  editBook: PropTypes.func.isRequired,
  removeBook: PropTypes.func.isRequired,
};

export default Book;
