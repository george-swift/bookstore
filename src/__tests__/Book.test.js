import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import '@testing-library/jest-dom/extend-expect';
import { render, within } from '@testing-library/react';
import Book from '../components/Book';

describe('Testing the Book Component', () => {
  const book = {
    id: 101,
    title: 'Test Book',
    category: 'Learning',
  };

  const editBook = jest.fn(() => {});
  const removeBook = jest.fn(() => {});

  beforeEach(() => render(
    <Router>
      <Book book={book} editBook={editBook} removeBook={removeBook} />
    </Router>,
  ));

  it('should render correctly', () => {
    const tree = renderer
      .create(
        <Router>
          <Book book={book} editBook={editBook} removeBook={removeBook} />
        </Router>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should have book title in info section', () => {
    const bookInfo = document.querySelector('.details');
    const testBookTitle = within(bookInfo).getByText('Test Book');
    expect(testBookTitle).toBeInTheDocument();
  });

  it('should have book category in info section', () => {
    const bookInfo = document.querySelector('.details');
    const testBookCategory = within(bookInfo).getByText('Learning');
    expect(testBookCategory).toBeInTheDocument();
  });

  it('should have author placeholder in info section by default', () => {
    const bookInfo = document.querySelector('.details');
    const placeholder = within(bookInfo).getByText('Click Edit to add author');
    expect(placeholder).toBeInTheDocument();
  });

  it('should have action button for comments', () => {
    const bookInfo = document.querySelector('.action');
    const comments = within(bookInfo).getByRole('button');
    expect(comments).toHaveTextContent('Comments');
  });

  it('should have action button for deleting a book', () => {
    const removeButton = document.querySelector('.remove');
    expect(removeButton).toHaveAttribute('title', 'Click to remove book from store');
    const action = within(removeButton).getByText('Remove');
    expect(action).toBeInTheDocument();
  });

  it('should have default value Zero for book progress', () => {
    const completed = document.querySelector('.percent');
    expect(completed).toHaveTextContent('0%');
    const currentChapter = document.querySelector('.current-lesson');
    expect(currentChapter).toHaveTextContent('Chapter 0');
  });

  it('should link to route for editing a book', () => {
    const editLink = document.querySelector('.edit');
    const action = within(editLink).getByText('Edit');
    expect(action).toBeInTheDocument();
    expect(action).toHaveAttribute('href', '/edit');
    const updateProgress = document.querySelector('.update-progress');
    const similarAction = within(updateProgress).getByText('Update Progress');
    expect(similarAction).toBeInTheDocument();
    expect(similarAction).toHaveAttribute('href', '/edit');
  });
});
