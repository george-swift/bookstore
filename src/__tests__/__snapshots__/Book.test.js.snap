// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the Book Component should render correctly 1`] = `
<div
  className="col-12 d-flex book-row"
>
  <div
    className="book-name"
  >
    <div
      className="details"
    >
      <p
        className="book-category"
      >
        Learning
      </p>
      <p
        className="title"
      >
        Test Book
      </p>
      <span
        className="d-block author"
      >
        Click Edit to add author
      </span>
      <ul>
        <li
          className="action border-end"
        >
          <span
            className="comments"
            role="button"
          >
            Comments
          </span>
        </li>
        <li
          className="action border-end"
        >
          <button
            className="remove"
            onClick={[Function]}
            title="Click to remove book from store"
            type="button"
          >
            Remove
          </button>
        </li>
        <li
          className="action"
        >
          <a
            className="edit"
            href="/edit"
            onClick={[Function]}
          >
            Edit
          </a>
        </li>
      </ul>
    </div>
    <div
      className="progress-ui"
    >
      <span
        className="progress-icon"
      >
        <svg
          className="CircularProgressbar "
          data-test-id="CircularProgressbar"
          style={Object {}}
          viewBox="0 0 100 100"
        >
          <path
            className="CircularProgressbar-trail"
            d="
      M 50,50
      m 0,-46
      a 46,46 0 1 1 0,92
      a 46,46 0 1 1 0,-92
    "
            fillOpacity={0}
            strokeWidth={8}
            style={
              Object {
                "strokeDasharray": "289.02652413026095px 289.02652413026095px",
                "strokeDashoffset": "0px",
              }
            }
          />
          <path
            className="CircularProgressbar-path"
            d="
      M 50,50
      m 0,-46
      a 46,46 0 1 1 0,92
      a 46,46 0 1 1 0,-92
    "
            fillOpacity={0}
            strokeWidth={8}
            style={
              Object {
                "strokeDasharray": "289.02652413026095px 289.02652413026095px",
                "strokeDashoffset": "289.02652413026095px",
              }
            }
          />
        </svg>
      </span>
      <p>
        <span
          className="percent"
        >
          0%
        </span>
        <br />
        <span
          className="completed"
        >
          Completed
        </span>
      </p>
    </div>
  </div>
  <span
    className="divider"
  />
  <div
    className="book-chapter"
  >
    <p
      className="current-chapter"
    >
      Current Chapter
    </p>
    <p
      className="current-lesson"
    >
      Chapter 0
    </p>
    <a
      className="btn btn-sm update-progress"
      href="/edit"
      onClick={[Function]}
    >
      Update Progress
    </a>
  </div>
</div>
`;
